pkg/client/bucket/bucket.go

```diff
@@ -0,0 +1,210 @@
+/*
+/*
+ * @license
+ * Copyright 2023 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package bucket
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/concurrency"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/rest"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const endpoint = "platform/storage/management/v1/bucket-definitions"
+
+type (
+	// Response holds all necessary information to create and update Grail Buckets
+	Response struct {
+		BucketName string `json:"bucketName"`
+		Status     string `json:"status"`
+		Version    int    `json:"version"`
+		Data       []byte `json:"-"`
+	}
+
+	// Client abstracts the API access for Grail Buckets
+	Client struct {
+		url     string
+		client  *rest.Client
+		limiter *concurrency.Limiter
+	}
+)
+
+// NewClient creates a new client to interact with the Grail Bucket API
+func NewClient(url string, client *rest.Client) *Client {
+	return &Client{
+		url:     url,
+		client:  client,
+		limiter: concurrency.NewLimiter(5),
+	}
+}
+
+// Upsert create or updates a given Grail Bucket
+func (c Client) Upsert(ctx context.Context, id string, data []byte) (result Response, err error) {
+	if id == "" {
+		return Response{}, fmt.Errorf("id must be non empty")
+	}
+	c.limiter.ExecuteBlocking(func() {
+		result, err = c.upsert(ctx, id, data)
+	})
+	return
+}
+
+// upsert first tries to create the bucket. If it does not work, we try to update it.
+// This has the advantage that we don't need to try to fetch the object first.
```

Comment:
This saves us 1 API call in the create case but potentially triggers unnecessary work on the Dynatrace environment. 
Generally fine with it, but I'm not sure if saving us one GET is worth the potential processing impact on Dynatrace. (We could check with GET /{id} in this case, right? we don't need to list all?)

Is the intention to keep this for the final version as well?

Reply:
With a proper list, we could save on requests (not data). However, with a simple Get, we ran into concurrency issues already in our very basic integration test: The object was created, but the GET did fail as the server said 'does not exist' as it has not yet 'recognized' the object; thus the update failed, as we expected to create an object. That's why we also have the retry in place. 
However, with the retry, the initial create took multiple requests, as the object did not exist in the first place (of cause, has not been created yet). With many objects, this really becomes an issue.

So actually, currently, without the list, we do safe on requests. Yes, the create could be more nice, but overall we do correctly create the object and don't need to execute multiple GET requests in order to do so.